### DISCUSSION TITLE
Enabling key modifiers for vue.js events

### DISCUSSION THREAD

**Discussion by cardinalgeo** [Kitware/trame#287] [Q&A] [answered]
Hi Trame team! 

`Vue.js` events are subject to key modifiers. For example, `@click.ctrl` and `@click.shift` are two variations of the click event that are triggered only when their respective keys are also pressed (e.g., as documented [here](https://vuejs.org/guide/essentials/event-handling.html#key-modifiers)).  This is useful for extending the functionality of a widget (e.g., a button that runs one function when clicked and another when clicked with the shift key down). However, it appears that this functionality is absent from `trame`. In the case of a `VBtn` from `trame-vuetify`, `VBtn(click=print_hello)` runs the function `print_hello` on clicking the button. However, `VBtn(click_shift=print_hello)` (or any variation) does not trigger an event when the button is clicked with the shift key pressed. This is consistent with the apparent absence of any reference to key modifiers in the documentation. 

Is it the case that key modifier functionality is not present? If so, what steps would be necessary to enable it? 

Thanks, Robert

**Comment by jourdain** [comment, marked as answer]
So in general we should improve the attribute management within trame to natively support what you are describing. 

But by nature trame already support everything that vue has to offer. 

The issue you are running into is that we did not already exposed that event. But you can [do whatever you want](https://trame.readthedocs.io/en/latest/core.widget.html) via the `__properties=[]` and `__events=[]`.

So to solve your use-case, you can do the following 
```python
VBtn(
   click_shift=print_hello, 
   __events=[
      ("click_shift", "click.shift"),
   ],
)
``` 
or even 
```python
VBtn(ab=print_hello, __events=[("ab", "click.shift")])
```

And as usual, if you want to double check that trame is generating the proper vue template, you can run the following code.

```python
btn = VBtn(click_shift=print_hello)
print(btn.html)
```

**Reply by cardinalgeo** [reply]
Per usual, thanks for the quick resolution to my problem! The flexibility enabled by `__events` and `__properties` will hopefully help me solve similar issues going forward. 

Of note for the solution above, while clicking the button triggers the `click` event, clicking the button with shift triggers **both** the `click` event and `click_shift` event. At first this was surprising (I've never used vue.js on its own), but I now see that it's probably the native functionality and I can easily get around the issue for my use case. Just documenting the behavior in case its useful for anyone else. (Also, I'd mark your answer as the solution, but that does not appear to be possible in the "general" category for discussions. )

Thanks again!

**Comment by jourdain** [comment]
Another thing that you can do is process the general click and decide what to do based on the [event](https://www.w3schools.com/jsref/obj_mouseevent.asp)

```python
VBtn(click=(print_hello, "[$event.buttons, $event.altKey, $event.ctrlKey, $event.metaKey, $event.shiftKey]")
```
